R/drive_download.R

```diff
@@ -43,49 +43,104 @@ drive_download <- function(file = NULL,
 
-
-  ext <- tools::file_ext(out_path)
-  if (ext == "" && !is.null(type)) {
-    out_path <- paste0(out_path, ".", type)
+  if (!grepl("google", mime_type) && !is.null(type)) {
+    message("Ignoring `type`. Only consulted for native Google file types.")
   }
 
   if (grepl("google", mime_type)) {
-    ## TODO check that extension is actually allowed per
-    ## https://developers.google.com/drive/v3/web/manage-downloads
-    ext <- tools::file_ext(out_path)
-    if (ext == "") {
-      stop(
-        glue(
-          "We don't know how to save the Google file: {sq(out_path)} \nPlease specify a file name with an exension in `out_path`."
-        )
+    export_type <- type %||% ext %||% get_export_mime_type(mime_type)
+    export_type <- drive_mime_type(export_type)
+    verify_export_mime_type(mime_type, export_type)
+    out_path <- apply_extension(out_path, export_type)
+
+    request <- generate_request(
+      endpoint = "drive.files.export",
+      params = list(
+        fileId = file$id,
+        mimeType = export_type
       )
-    }
-    mime_type <- drive_mime_type(ext)
-    request <- generate_request(endpoint = "drive.files.export",
-                                params = list(fileId = file$id,
-                                              mimeType = mime_type ))
+    )
   } else {
-    request <- generate_request(endpoint = "drive.files.get",
-                                params = list(alt = "media",
-                                              fileId = file$id))
+    request <- generate_request(
+      endpoint = "drive.files.get",
+      params = list(
+        fileId = file$id,
+        alt = "media"
+      )
+    )
   }
 
-  response <- make_request(request, httr::write_disk(out_path, overwrite = overwrite))
-  success <- file.exists(out_path)
+  response <- make_request(
+    request,
+    httr::write_disk(out_path, overwrite = overwrite)
+  )
+  success <- httr::status_code(response) == 200 && file.exists(out_path)
 
   if (success) {
     if (verbose) {
       message(
-        glue("File downloaded from Google Drive:\n{sq(file$name)}\nSaved locally as:\n{sq(out_path)}")
+        glue(
+          "File downloaded from Google Drive:\n{sq(file$name)}\n",
+          "Saved locally as:\n{sq(out_path)}"
+        )
       )
     }
   } else {
-    spf("Zoinks! the file doesn't seem to have downloaded")
+    stop("Zoinks! the file doesn't seem to have downloaded", call. = FALSE)
   }
   invisible(file)
 }
 
+## get the default export MIME type for a native Google MIME type
+## examples:
+##    Google Doc --> MS Word
+##  Google Sheet --> MS Excel
+## Google Slides --> MS PowerPoint
+get_export_mime_type <- function(mime_type) {
+  m <- .drive$translate_mime_types$mime_type_google == mime_type &
+    is_true(.drive$translate_mime_types$default)
+  if (!any(m)) {
+    stop(glue("Not a recognized Google MIME type:\n{mime_type}"), call. = FALSE)
+  }
+  .drive$translate_mime_types$mime_type_local[m]
+}
```

Comment:
If we keep this, we need to make it easier for people to see what file types are available for each Google type and what our chosen default is.